cpp/src/DataStorm/NodeI.cpp

```diff
@@ -373,50 +403,73 @@ NodeI::createPublisherSession(
             false,
             nullptr,
             [self = shared_from_this(), publisher, session](exception_ptr ex)
-            { self->removeSubscriberSession(publisher, session, ex); });
+            { self->retrySubscriberSessionCreation(publisher, session, ex); });
     }
     catch (const LocalException&)
     {
-        removeSubscriberSession(publisher, session, current_exception());
+        if (session)
+        {
+            retrySubscriberSessionCreation(publisher, session, current_exception());
+        }
+        // Else node is shutting down.
+        throw SessionCreationException{SessionError::Internal};
     }
 }
 
 void
-NodeI::removeSubscriberSession(const NodePrx& node, const shared_ptr<SubscriberSessionI>& session, exception_ptr ex)
+NodeI::retrySubscriberSessionCreation(
```

Comment:
I split removeSubscriberSession into two methods: retrySubscriberSessionCreation and removeSubscriberSession. The removeSubscriberSession method is now only called when a retry isn’t possible. I think this separation makes the intent clearer.